File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        let length = nums.length;
        let left = new Array(length).fill(1);
        let right = new Array(length).fill(1);
        for(let i=1; i<length;i++){
            left[i] = left[i-1]*nums[i-1];
        }
        for(let i=length-2; i>=0; i--){
            right[i] = right[i+1]*nums[i+1];
        }
        let ans = [];
        for(let i=0; i<length; i++){
            ans.push(left[i]*right[i]);
        }
        return ans;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-2.java
class Solution {
    public int trap(int[] height) {
        int[]left = new int[height.length];
        int[]right = new int[height.length];
        for(int i=1;i<height.length;i++){
            left[i]=Math.max(left[i-1],height[i-1]);
            right[height.length-1-i]=Math.max(right[height.length-i],height[height.length-i]);
        }
        int ans = 0;
        for(int i=0;i<height.length;i++){
            ans += Math.min(left[i],right[i])-height[i]>0?Math.min(left[i],right[i])-height[i]:0;
        }
        return ans;  
    }
}
